src/Unix-Timestamp/unixts2time.1
.\" ==========================================================================
.\"         ____            _                     _____           _
.\"        / ___| _   _ ___| |_ ___ _ __ ___     |_   _|__   ___ | |___
.\"        \___ \| | | / __| __/ _ \ '_ ` _ \ _____| |/ _ \ / _ \| / __|
.\"         ___) | |_| \__ \ ||  __/ | | | | |_____| | (_) | (_) | \__ \
.\"        |____/ \__, |___/\__\___|_| |_| |_|     |_|\___/ \___/|_|___/
.\"               |___/
.\"                             --- System-Tools ---
.\"                  https://www.nntb.no/~dreibh/system-tools/
.\" ==========================================================================
.\"
.\" Unix Timestamp to Time Conversion
.\" Copyright (C) 2026 by Thomas Dreibholz
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.\" Contact: thomas.dreibholz@gmail.com
.\"
.\" ###### Setup ############################################################
.Dd September 14, 2026
.Dt unixts2time 1
.Os unixts2time
.\" ###### Name #############################################################
.Sh NAME
.Nm unixts2time
.Nd converts Unix timestamps to date/time strings
.\" ###### Synopsis #########################################################
.\" Manpage syntax help:
.\" https://forums.freebsd.org/threads/howto-create-a-manpage-from-scratch.13200/
.Sh SYNOPSIS
.Nm unixts2time
.Op Fl T Ar time\_format\_template | Fl Fl template Ar time\_format\_template
.br
.Op Fl H | Fl Fl human\-readable
.br
.Op Fl s | Fl Fl seconds
.Op Fl m | Fl Fl milliseconds
.Op Fl u | Fl Fl microseconds
.Op Fl n | Fl Fl nanoseconds
.br
unix\_timestamp
.Op ...
.Nm unixts2time
.Op Fl h | Fl Fl help
.Nm unixts2time
.Op Fl v | Fl Fl version
.\" ###### Description ######################################################
.Sh DESCRIPTION
.Nm unixts2time
converts Unix timestamps to date/time strings in the format %Y\-%m\-%d %H:%M:%S.SSSSSSSSS. A Unix timestamp is the time (nanoseconds, microseconds, milliseconds, or seconds) since the Unix Epoch, i.e. January 1, 1970, 00:00:00.000000000 UTC.
.br
By default, the timestamp-granularity is auto\-detected using the following heuristic:
.Bl -bullet -offset indent
.It
Seconds (s): from
0 (1970\-01\-01 00:00:00) to 4999999999 = 5e9\-1 (2128\-06\-11 08:53:19), and
\-4999999999 = \-5e9 (1811\-07\-23 15:06:41) to \-1 (1969\-12\-31 23:59:59);
.It
Milliseconds (ms): from
5000000000 = 5e9 (1970\-02\-27 20:53:20.000) to 4999999999999 = 5e12\-1 (2128\-06\-11 08:53:19.999) and
\-4999999999999 = \-5e12\-1 (1811\-07\-23 15:06:41.999) to \-5000000000 = \-5e9 (1969\-11\-04 03:06:40.000);
.It
Microseconds (µs) from
5000000000000 = 5e12 (1970\-02\-27 20:53:20.000000) to 4999999999999999 = 5e15\-1 (2128\-06\-11 08:53:19.999999) and
\-4999999999999999 = \-5e15\-1 (1811\-07\-23 15:06:41.999999) to \-5000000000000 = \-5e12 (1969\-11\-04 03:06:40.000000);
.It
Nanoseconds (ns) from
5000000000000000 = 5e15 (1970\-02\-27 20:53:20.000000000) and
\-5000000000000000 = \-5e15 (1969\-11\-04 03:06:40.000000000).
.El
That is, unless using dates around 1969/1970, ancient past, or distant future, the auto\-detection should work reasonably well. Optionally, the timestamp granularity can be specified explicitly.
.Pp
.\" ###### Arguments ########################################################
.Sh ARGUMENTS
The following options are available:
.Bl -tag -width indent
.It unix\_timestamp ...
The Unix timestamps to be converted. If no Unix timestamp is provided, the current system time is used. The timestamp may be provided as float or integer. For integers, the numbers can have a prefix 0x for hexadecimal (base 16), or 0 for octal (base 8). Otherwise, the number is interpreted as decimal (base 10).
.It Fl T Ar time\_format\_template | Fl Fl template Ar time\_format\_template
Sets the template for printing the date/time strings. Default is \(dq"%Y-%m-%d %H:%M:%S\(dq. The formatting is according to strftime(). See
.Xr strftime 3
for details.
.It Fl H | Fl Fl human\-readable
Provide the output in human\-readable form. Default: off.
.It Fl s | Fl Fl seconds
The Unix timestamp is given in seconds. Default is auto\-detection (see above).
.It Fl m | Fl Fl milliseconds
The Unix timestamp is given in milliseconds. Default is auto\-detection (see above).
.It Fl u | Fl Fl microseconds
The Unix timestamp is given in microseconds. Default is auto\-detection (see above).
.It Fl n | Fl Fl nanoseconds
The Unix timestamp is given in nanoseconds. Default is auto\-detection (see above).
.It Fl h | Fl Fl help
Prints the program's command\-line parameters and exits.
.It Fl v | Fl Fl version
Prints the program version and exits.
.El
.\" ###### Exit status ######################################################
.Sh EXIT STATUS
The
.Nm
tool exits with 0 on success, and >0 in case of an error.
.\" ###### Examples #########################################################
.Sh EXAMPLES
.Bl -tag -width indent
.It unixts2time
.It unixts2time \-\-human\-readable
.It unixts2time \-\-human\-readable \-\-seconds 1773744755
.It unixts2time \-\-human\-readable \-\-seconds \-\- \-1000000
.It unixts2time \-\-human\-readable \-\-milliseconds \-\- \-442460258000
.It unixts2time \-\-human\-readable \-\-seconds 1773736279
.It unixts2time \-\-seconds 1 1000 1000000 1000000000 2000000000
.It unixts2time \-s 1773736279
.It unixts2time \-\-milliseconds 212847000000
.It unixts2time \-H 212847008123456789
.It unixts2time 0x179cb2ae8af3dd89
.It unixts2time \-H 0x179cb2ae8af3dd89
.It unixts2time \-s 212847123.123
.It unixts2time 1788550214.7787085
.It unixts2time \-T \(dq%x %X\(dq \-s 2096218885000000000
.It unixts2time \-T \(dq%b %d %H:%M:%S %Y %Z\(dq 1773736279
.It unixts2time \-T \(dq%b %d %H:%M:%S %Y %Z\(dq 1784186916836355733
.It unixts2time \-T \(dq%x %X\(dq 1784186916836355733
.It unixts2time \-\-version
.It unixts2time \-\-help
.El
.\" ###### See also #########################################################
.Sh SEE ALSO
.Xr strftime 3
.Xr time2unixts 1
.\" ###### Notes ############################################################
.Sh NOTES
This program is part of System\-Tools. The latest version of System\-Tools can be found on the System\-Tools Homepage at
.Lk https://www.nntb.no/\(tidreibh/system\-tools/ "System\-Tools Homepage" .
.\" ###### Authors ##########################################################
.Sh AUTHORS
Thomas Dreibholz,
.Lk https://www.nntb.no/\(tidreibh "Homepage"


ChangeLog
commit 0273e844adfe18591857f25a700da1efafae3566
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 17:05:08 2026 +0200

    Updated PO files.

commit 40d566574253e104ef1d8262409528e73e13f313
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 17:03:38 2026 +0200

    Bugfix: system-info.d and system-maintenance.d under Homebrew contain symlinks, not files.
    
    Also improved handling of script directories.

commit 08003be7000f240929b940964708ab664964dc6f
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 16:08:34 2026 +0200

    Updated FreeBSD port.

commit 8afc7ab571b6d051f303bb02bdafd55c5e08a03c
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 12:25:01 2026 +0200

    Updated APKBUILD file.

commit 5d2eb9cc96252ebc46630d39f9e3fd6c4dff49ab
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Sep 4 18:22:22 2026 +0200

    Updated APKBUILD file.

commit 587ce77acef704b83f6d1452451b0100f036da9b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Sep 4 18:13:14 2026 +0200

    New release td-system-tools-2.7.13.

commit 8d79add36e806c3976b3860e870e3bec61681aae
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Sep 4 18:09:04 2026 +0200

    PO file updates.

commit 700ba8a7c02336395182b8f2004a206dd990e29e
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Sep 2 21:28:51 2026 +0200

    Updated FreeBSD port.

commit ab19df84aaf7d7e7a7f8a90faad8e7162ef0de11
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 16:23:43 2026 +0200

    Updated APKBUILD file.

commit 46ec3f27549ea59e0a5ee86e3ded61ca8a367eb7
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Aug 31 16:27:43 2026 +0200

    New release td-system-tools-2.7.12.

commit 0b219ca51062396c85d114933be2e6d0361fe1f8
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Aug 27 09:28:47 2026 +0000

    Improved SSHFP RR output.

commit c3c0f692061daee599598b41d1244e70f8890d06
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Aug 27 09:00:11 2026 +0000

    Bugfix: Fixed querying the swap size under Solaris.

commit 54d2dbba2c6d4f1ba16b8ddf21a76356750f719f
Merge: 2e48c7d f7d36f8
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Aug 11 14:52:15 2026 +0000

    WIP on master: 2e48c7d Updated FreeBSD port.

commit 2e48c7d12ac602575a36ce4d19e53f700de8b92f
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Jul 31 13:49:50 2026 +0200

    Updated FreeBSD port.

commit 2d15c6435117cd7344777bb282d09f30481b08da
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Jul 31 13:21:42 2026 +0200

    Updated APKBUILD file.

commit 7a87b24841deddc5c6d5c27f9718001667375054
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Jul 31 12:49:30 2026 +0200

    New release td-system-tools-2.7.11.

commit 38752f318e2dec60f918678400c8cb7755e042eb
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Jul 31 11:26:42 2026 +0200

    Further clean-ups.

commit 776e793eca014c49a661105c628b399fd203b235
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Jul 31 11:24:23 2026 +0200

    Further ShellCheck clean-ups.

commit 660a4174809a25da69ad67ca390466e376cc1d53
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Jul 31 11:19:56 2026 +0200

    Minor clean-up.

commit 96983e49205334ba499846d12cff8a7675d1ca92
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Jul 31 11:19:06 2026 +0200

    Improved compatibility of ShellCheck script.

commit 4b779c0d543f127805a5cc7f95e7ffd21d6f78f9
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Jul 28 15:11:18 2026 +0200

    Updated PO files.

commit ef723836537074580f41524970d0c76358e18893
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Jul 28 13:08:31 2026 +0000

    Solaris UTF-8 character width computation fix.

commit 5ba247e608aae3bbb0badde0866172f113a82da4
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Jul 28 14:34:46 2026 +0200

    Updated translations.

commit 23f14ac5698a53e8a330412b7a4a52739c259c7c
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Jul 28 14:28:12 2026 +0200

    Improved checks for Bash version and enhanced getopt.

commit 0e00b3fb1e1323fdeaac952cd9a5c522c1c109df
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Jul 28 12:58:39 2026 +0200

    Also need to set TEXTDOMAINDIR in scripts explicitly.

commit 39a99cfe9e7669d92e62245a8a66629a64b8e567
Merge: 727a9c6 d13052b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Jul 28 12:42:43 2026 +0200

    Merge branch 'master' of github.com:dreibh/system-tools

commit 727a9c6306d6073b41242b26511e344754a0a62f
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Jul 28 12:41:57 2026 +0200

    Explicitly set locale directory in bindtextdomain() call.

commit d13052bb7adb6d57d5fc94a246932d0da8100f9c
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Jul 27 20:49:26 2026 +0000

    Updated comment.

commit 354674a0425a25975e2ccfb6dfa8e408dd705c98
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Jul 27 20:46:58 2026 +0000

    Added Solaris battery status block.

commit 7394954f5f0000425372b73f736e9711fede8b04
Merge: 8d67bd5 825fcf2
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Jul 27 15:45:14 2026 +0200

    Merge branch 'master' of github.com:dreibh/system-tools

commit 8d67bd5f749c54f04b5f2dddf6f2d8799f7a7565
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Jul 27 15:45:00 2026 +0200

    Improved example.

commit c656075f9f9d427c81b2ef66648436c53ebd3361
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Jul 27 15:44:22 2026 +0200

    Added test script.

commit 825fcf2551699165b552a0d873f8e1e797ff2d8a
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Jul 24 21:21:11 2026 +0000

    Fixed Debian/armhf build issue.

commit 3acd852be15d3d35f2e9971981e14ea106418e14
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Jul 24 22:36:45 2026 +0200

    CMakeLists.txt update to use C23.

commit 99d67b708afe8e44a8edbf119aae5e2b46045aae

src/Unix-Timestamp/unixts2time.c
// ==========================================================================
//         ____            _                     _____           _
//        / ___| _   _ ___| |_ ___ _ __ ___     |_   _|__   ___ | |___
//        \___ \| | | / __| __/ _ \ '_ ` _ \ _____| |/ _ \ / _ \| / __|
//         ___) | |_| \__ \ ||  __/ | | | | |_____| | (_) | (_) | \__ \.
//        |____/ \__, |___/\__\___|_| |_| |_|     |_|\___/ \___/|_|___/
//               |___/
//                             --- System-Tools ---
//                  https://www.nntb.no/~dreibh/system-tools/
// ==========================================================================
//
// Unix Timestamp to Time Conversion
// Copyright (C) 2026 by Thomas Dreibholz
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
// Contact: thomas.dreibholz@gmail.com

#define _GNU_SOURCE
#define __EXTENSIONS__
#include <ctype.h>
#include <errno.h>
#include <getopt.h>
#include <locale.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#ifdef ENABLE_NLS
#include <libintl.h>
#else
#define bindtextdomain(domain, dirname) { }
#define textdomain(domain) { }
#define gettext(string) string
#define ngettext(singular, plural, n) ((n) == 1 ? (singular) : (plural))
#endif

#include "package-version.h"

#if defined(__STDC_VERSION__) && (__STDC_VERSION__ < 202311L)
#ifndef nullptr
#define nullptr ((void*)0)
#endif
#endif


// ###### Find %S placeholder for seconds in time format string #############
static const char* findSecondsPlaceholder(const char* formatString)
{
   const char* ptr = formatString;
   while(*ptr) {
      if(*ptr == '%') {
         switch(*(ptr + 1)) {
            case 0x00:
               ptr++;
             break;
            case 'S':
               return ptr;
             break;
            default:
               ptr += 2;
             break;
         }
      } else {
         ptr++;
      }
   }
   return nullptr;
}


// ###### Version ###########################################################
#if defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 202000L)
[[ noreturn ]]
#endif
static void version(void)
{
   printf("unixts2time %s\n", SYSTEMTOOLS_VERSION);
   exit(0);
}


// ###### Usage #############################################################
#if defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 202000L)
[[ noreturn ]]
#endif
static void usage(const char* program, const int exitCode)
{
   fprintf(stderr, "%s %s unix_timestamp [...]"
           " [-T|--template time_format_template]"
           " [-H|--human-readable]"
           " [-s|--seconds]"
           " [-m|--milliseconds]"
           " [-u|--microseconds]"
           " [-n|--nanoseconds]"
           " [-h|--help]"
           " [-v|--version]\n",
           gettext("Usage:"), program);
   exit(exitCode);
}



// ###### Main program ######################################################
int main(int argc, char** argv)
{
   // ====== Initialise i18n support ========================================
   if(setlocale(LC_ALL, "") == nullptr) {
      setlocale(LC_ALL, "C.UTF-8");   // "C" should exist on all systems!
   }
   setlocale(LC_NUMERIC, "C");   // Use "." for fractional numbers!
   bindtextdomain("unixts2time", SYSTEMTOOLS_LOCALEDIR);
   textdomain("unixts2time");

   // ====== Handle arguments ===============================================
   static const struct option long_options[] = {
      { "template",       required_argument, 0, 'T' },
      { "human-readable", no_argument,       0, 'H' },
      { "seconds",        no_argument,       0, 's' },
      { "milliseconds",   no_argument,       0, 'm' },
      { "microseconds",   no_argument,       0, 'u' },
      { "nanoseconds",    no_argument,       0, 'n' },
      { "help",           no_argument,       0, 'h' },
      { "version",        no_argument,       0, 'v' },
      {  nullptr,         0,                 0, 0   }
   };

   int          option;
   int          longIndex;
   bool         humanReadable      = false;
   unsigned int initialDivideBy    = 0;         // auto-detect later
   const char*  unit               = nullptr;   // auto-detect later
   const char*  timeFormatTemplate = "%Y-%m-%d %H:%M:%S";
   while( (option = getopt_long(argc, argv, "T:Hsmunvh", long_options, &longIndex)) != -1 ) {
      switch(option) {
         case 'T':
            timeFormatTemplate = optarg;
          break;
         case 'H':
            humanReadable = true;
          break;
         case 'n':
            initialDivideBy = 1;
            unit     = "ns";
          break;
         case 'u':
            initialDivideBy = 1000;
            unit     = "µs";
          break;
         case 'm':
            initialDivideBy = 1000000;
            unit     = "ms";
          break;
         case 's':
            initialDivideBy = 1000000000;
            unit     = "s";
          break;
         case 'v':
            version();
          break;
         case 'h':
         case '?':
            // Exit with 0 on h/help, exit with 1 on '?' (unknown option):
            usage(argv[0], (option == 'h') ? 0 : 1);
          break;
         case '-':
          break;
         default:
            // This should not happen: wrong getopt parameters, or missing case?
            fprintf(stderr, "INTERNAL ERROR: Unhandled option c=%c code=%x!\n",
                    (isprint(option) ? (char)option : ' '), option);
            return 1;
          break;
      }
   }

   // ====== Obtain Unix timestamp in ns ====================================
#if (defined(__BITINT_MAXWIDTH__) && (__BITINT_MAXWIDTH__ >= 128))
   _BitInt(128) unixTS;
#else
   // NOTE: A 64-bit signed long long will overflow on April 11, 2262!
   long long    unixTS;
#endif
   for(int i = optind; i <= argc; i++) {
      unsigned int divideBy              = initialDivideBy;
      unsigned int inputFractionalDigits = 0;

      // ====== Use current time, if no timestamp is given ==================
      if(i == argc) {
         if(optind == argc) {
            struct timespec ts;
            if(clock_gettime(CLOCK_REALTIME, &ts) == -1) {
               perror(gettext("clock_gettime() failed"));
               exit(1);
            }
#if (defined(__BITINT_MAXWIDTH__) && (__BITINT_MAXWIDTH__ >= 128))
            unixTS = ((_BitInt(128))1000000000LL * ts.tv_sec) + ts.tv_nsec;
#else
            unixTS = (1000000000LL * ts.tv_sec) + ts.tv_nsec;
#endif
            if(divideBy == 0) {
               divideBy = 1;
               unit     = "ns";
            }
         }
         else {
            break;
         }
      }

      // ====== Parse the next Unix timestamp ===============================
      else {
         // ------ Try to parse integer part --------------------------------
         char* endptr;
         unixTS = strtoll(argv[i], &endptr, 0);
         if(endptr != nullptr) {
            if(divideBy == 0) {
               if( (unixTS > -5000000000LL) && (unixTS < 5000000000LL) ) {
                  divideBy = 1000000000;
                  unit     = "s";
               }
               else if( (unixTS > -5000000000000LL) && (unixTS < 5000000000000LL) ) {
                  divideBy = 1000000;
                  unit     = "ms";
               }
               else if( (unixTS > -5000000000000000LL) && (unixTS < 5000000000000000LL) ) {
                  divideBy = 1000;
                  unit     = "µs";
               }
               else {
                  divideBy = 1;
                  unit     = "ns";
               }
            }
            unixTS *= divideBy;   // convert to ns

            // ------ Try to parse fractional part --------------------------
            if(*endptr != 0x00) {
               if(*endptr == '.') {
                  const char* p = endptr + 1;
                  while((*p >= '0') && (*p <= '9')) {
                     inputFractionalDigits++;
                     p++;
                  }
               }

               const double fractionalUnixTS = strtod(endptr, &endptr);
               if(endptr != nullptr) {
                  const unsigned int additionalNS =
                     (unsigned int)(fractionalUnixTS * divideBy);
                  unixTS = (unixTS >= 0) ? (unixTS + additionalNS) :
                                           (unixTS - additionalNS);
               }
            }
         }
         if( (endptr == nullptr) || (*endptr != 0x00) ) {
            fputs(gettext("ERROR: Invalid Unix timestamp!"), stderr);
            fputs("\n", stderr);
            exit(1);
         }
      }

      // ====== Convert Unix timestamp to time ==============================
      struct timespec ts = {
         unixTS / 1000000000,
         unixTS % 1000000000
      };
      if( (unixTS < 0) && (ts.tv_nsec < 0) ) {
         ts.tv_sec--;
         ts.tv_nsec = 1000000000 + ts.tv_nsec;
      }

      // ------ Time in seconds-granularity ---------------------------------
      const struct tm* t = gmtime(&ts.tv_sec);
      if(t == nullptr) {
         fputs(gettext("ERROR: gmtime() failed!"), stderr);
         fputs("\n", stderr);
         exit(1);
      }

      // ====== Prepare date/time string ====================================
      char frontTimeString[1024]        = { 0 };
      char secondsString[128]           = { 0 };
      char fractionalSecondsString[128] = { 0 };
      char backTimeString[1024]         = { 0 };

      const char* secondsPlaceholder = findSecondsPlaceholder(timeFormatTemplate);
      if(secondsPlaceholder != nullptr) {
         // ------ Prepare front part (before seconds) ----------------------
         const size_t frontLength = (size_t)(secondsPlaceholder - timeFormatTemplate);
         char* frontFormatString  = malloc(frontLength + 1);
         if(frontFormatString == nullptr) {
            fprintf(stderr, gettext("ERROR: malloc() failed: %s!"), strerror(errno));
            exit(1);
         }
         memcpy(frontFormatString, timeFormatTemplate, frontLength);
         frontFormatString[frontLength] = 0x00;
         if(strftime(frontTimeString, sizeof(frontTimeString), frontFormatString, t) == 0) {
            fprintf(stderr, gettext("ERROR: strftime() with format template \"%s\" failed!"),
                    frontFormatString);
            fputs("\n", stderr);
            free(frontFormatString);
            exit(1);
         }
         free(frontFormatString);

         // ------ Parse the middle part (seconds) --------------------------
         if(strftime(secondsString, sizeof(secondsString), "%S", t) == 0) {
            fprintf(stderr, gettext("ERROR: strftime() with format template \"%s\" failed!"),
                    "%S");
            fputs("\n", stderr);
            exit(1);
         }

         // ------ Prepare the fractional seconds string --------------------
         unsigned int precision = 0;
         if(divideBy == 1) {
            precision = 9;
         }
         else if(divideBy == 1000) {
            precision = 6;
         }
         else if(divideBy == 1000000) {
           precision = 3;
         }
         precision += inputFractionalDigits;
         if(precision > 9) {
            precision = 9;
         }

         char fractionalSecondsFormatString[32];
         snprintf(fractionalSecondsFormatString, sizeof(fractionalSecondsFormatString),
                  "%%1.%uf", precision);
         snprintf(fractionalSecondsString, sizeof(fractionalSecondsString),
                  fractionalSecondsFormatString,
                  (double)ts.tv_nsec / 1000000000.0);

         // ------ Prepare the back part (after seconds) --------------------
         const char* backFormatString = secondsPlaceholder + 2;
         if(backFormatString[0] != 0x00) {
            if(strftime(backTimeString, sizeof(backTimeString), backFormatString, t) == 0) {
               fprintf(stderr, gettext("ERROR: strftime() with format template \"%s\" failed!"),
                       backFormatString);
               fputs("\n", stderr);
               exit(1);
            }
         }
      }
      else {
         if(strftime(frontTimeString, sizeof(frontTimeString), timeFormatTemplate, t) == 0) {
            fprintf(stderr, gettext("ERROR: strftime() with format template \"%s\" failed!"),
                    timeFormatTemplate);
            fputs("\n", stderr);
            exit(1);
         }
      }

      // ====== Print result ================================================
      if(!humanReadable) {
         fputs(frontTimeString, stdout);
         fputs(secondsString, stdout);
         if(fractionalSecondsString[0] != 0x00) {
            fputs(&fractionalSecondsString[1], stdout);
         }
         fputs(backTimeString, stdout);
      }
      else {
         fprintf(stdout, gettext("%s%s%s%s is %lld (0x%llx) %s since the Unix Epoch"),
                 frontTimeString,
                 secondsString, (fractionalSecondsString[0] != 0x00) ? fractionalSecondsString + 1 : "",
                 backTimeString,
                 (long long)(unixTS / divideBy),
                 (unsigned long long)(unixTS / divideBy),
                 unit);
      }
      fputs("\n", stdout);
   }

   return 0;
}
